### PR TITLE
Reserve guarded string literals (RFC 3593)

### PR DESCRIPTION
**Pull request by pitaj**
Implementation for RFC 3593, including:
- lexer / parser changes
- diagnostics
- migration lint
- tests

We reserve `#"`, `##"`, `###"`, `####`, and any other string of four or more repeated `#`. This avoids infinite lookahead in the lexer, though we still use infinite lookahead in the parser to provide better forward compatibility diagnostics.

This PR does not implement any special lexing of the string internals:
- strings preceded by one or more `#` are denied
- regardless of the number of trailing `#`
- string contents are lexed as if it was just a bare `"string"`

Tracking issue: #123735
RFC: rust-lang/rfcs#3593

### CODE REVIEW

**Comment by mattheww** [comment]
In edition 2024 I'm seeing a panic inside `cook_unicode` if I give this an empty guarded string without enough closing hashes (eg `#""` or `##""#`).

**Comment by pitaj** [comment]
@mattheww good catch with the empty strings. I wasn't able to reproduce the "double-warning" issue with `blah"xx"`, but just managed to do so with `blah#"xx"#`. 

I think it's actually the correct behavior. When you fix the issue with `blah #"xx"#` as recommended, you'll end up getting the second error anyways. Better to get both at once, then you can decide where to put the space.

Regardless, that case will be exceedingly rare if it exists in the wild at all, so I'm not going to spend any time on it.

**Comment by mattheww** [comment]
My stress-tester is now giving an ICE for `£#""#` in editions older than 2024.

File: tests/ui/rust-2024/reserved-guarded-strings.stderr
Comment (estebank) [inline tests/ui/rust-2024/reserved-guarded-strings.stderr:131]: This is less than ideal. We'd want to suggest `r##` for these with the right number of `#` at both ends instead.

Comment (pitaj) [reply tests/ui/rust-2024/reserved-guarded-strings.stderr:131]: That's a good point. Though I don't think this is a downgrade from the current behavior:
```rust
error: expected one of `!` or `[`, found `#`
 --> src/main.rs:2:14
  |
2 |     let s = ##"blah"##;
  |              ^ expected one of `!` or `[`
```

**Comment by mattheww** [comment]
My lexer stress-tests don't find any problems with this now.

(They're checking for ICEs, cases where the lexer output changes in previous editions, and cases where the lexer output is not as expected in the 2024 edition.)

**Comment by pitaj** [comment]
The crater run for #124605 [found some real regressions](https://crater-reports.s3.amazonaws.com/pr-124605/index.html). So that proves the effort made here for backwards compatibility is warranted.

File: compiler/rustc_lexer/src/lib.rs
Comment (estebank) [inline compiler/rustc_lexer/src/lib.rs:377]: So far, this approach has *only* being used for the benefit of diagnostics, but won't this become a language requirement to have this kind of "unlimited" lookahead in all alternative implementations?

Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs:377]: It depends. Alternative implementations could choose to have fully edition-dependent lexing. In that case, they would not need this backtracking.

**Comment by traviscross** [comment]
We reviewed this in the edition call today.  What's the next step here, e.g. is this waiting on further review or on updates from @pitaj?

cc @estebank

**Comment by estebank** [comment]
@traviscross I'm just concerned with introducing the snapshotting behavior into the lexer. There's no precedent for that approach in that stage of the compiler. It might be that there's no alternative, but would like more people in @rust-lang/compiler to double check on the implementation.

**Comment by oli-obk** [comment]
cc @Nilstrieb @compiler-errors fun lexer things

**Comment by Nilstrieb** [comment]
forwarding my ping to @nnethercote

**Comment by jackh726** [comment]
Maybe we just compiler-nominate this?

**Comment by nnethercote** [comment]
I can take a look, I'm pretty familiar with this code.

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_lexer/src/lib.rs:184]: Two thumbs up for fixing these quotes.

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_lexer/src/lib.rs:200]: This confuses me.
- "`None` indicates no closing quote." Wouldn't it make sense for `n_end_hashes` to have the `Option` instead of `n_start_hashes`?
- "Allows fewer hashes to close the string to support older editions."  What does this mean? This is a new feature, how are older editions relevant? Why is an unequal number of hashes allowed in older editions?

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_lexer/src/lib.rs:825]: The example `GuardedStr` has one field, but the type has two fields. So the example needs updating.

File: compiler/rustc_lint/src/context/diagnostics.rs
Comment (nnethercote) [inline compiler/rustc_lint/src/context/diagnostics.rs:164]:
```suggestion
                    "insert whitespace here to avoid this being parsed as a guarded string in Rust 2024",
```

File: compiler/rustc_lint/src/context/diagnostics.rs
Comment (nnethercote) [inline compiler/rustc_lint/src/context/diagnostics.rs:170]:
```suggestion
                    "insert whitespace between the `#`s and the opening quote to avoid this being parsed as a guarded string in Rust 2024",
```

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_lexer/src/lib.rs:374]: I feel like most of this match arm should be moved into a separate function, as is done for cases like `c_or_byte_string`.

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_lexer/src/lib.rs:833]: I can't see any checking that the number of start and end hashes match. Am I overlooking it?

File: compiler/rustc_lexer/src/lib.rs
Comment (nnethercote) [reply compiler/rustc_lexer/src/lib.rs:377]: I also don't like this snapshotting. Should `Cursor` gain an `edition` field? Seems like the right way to do it.

**Comment by nnethercote** [comment]
This was the first I'd heard of this feature. At the moment, I am not feeling positive about it :(
- Yet another, slightly different string literal kind? When will they stop?
- Why are they called "guarded string literals"? The "guarded" doesn't mean anything to me.
- The RFC says they unify string literals and raw string literals. I understand how guarded string literals combine characteristics of string literals and raw string literals, but don't see how adding a new string literal "unifies" anything. At best guarded string literals would replace raw string literals (they are strictly more powerful) but that seems very difficult to do due to backwards compatibility concerns.
- Starting with `#` complicates the lexer quite a lot. An alternative that the RFC didn't mention would be to use a different identifying letter, such as `g`. E.g. `g#"foo"#`. I think that would avoid a lot of the complexity, and could be introduced more easily, as was done with `c` string literals.
- That wouldn't match what Swift does, but is that a problem? If we were redesigning Rust from scratch then maybe guarded literals would make sense, but with the history and backwards compatibility story of Rust, and the existing string literals forms, I'm not so sure.

I don't like to be so negative, but I'm quite familiar with the relevant code in the lexer and parser (e.g. from working on [RFC3349](https://github.com/rust-lang/rfcs/pull/3349)) and the whole proposal just doesn't feel good to me :(

**Comment by pitaj** [comment]
@nnethercote did you mean to comment this on https://github.com/rust-lang/rfcs/pull/3475 ?

**Comment by pitaj** [comment]
Okay rebased and fixed a few minor issues. To address your questions:

> The RFC says they unify string literals and raw string literals. I understand how guarded string literals combine characteristics of string literals and raw string literals, but don't see how adding a new string literal "unifies" anything.

The proposed "unified" string literals are an extension of normal string literals, with the capabilities of raw string literals. Retroactively, every normal string literal is a guarded string literal with zero wrapping `#`. They merge the capabilities of raw string literals into a unified syntax with normal string literals.

> Yet another, slightly different string literal kind? When will they stop?

> At best guarded string literals would replace raw string literals (they are strictly more powerful) but that seems very difficult to do due to backwards compatibility concerns.

Yes, they are meant to replace raw string literals. Both would exist for one edition cycle, and at the following edition boundary, all raw string literals would be migrated to guarded string literals. I don't see why there would be backwards compatibility issues with that path.

> Starting with `#` complicates the lexer quite a lot. An alternative that the RFC didn't mention would be to use a different identifying letter, such as `g`. E.g. `g#"foo"#`. I think that would avoid a lot of the complexity, and could be introduced more easily, as was done with `c` string literals.

One of the great benefits of unified string literals is that they're completely composable with `c` strings, `b` strings, etc. No need for `b"foo"`, `br#"foo"#`, `c"foo"`, `cr#"foo"#`, etc. With unified string literals, you just have the prefix (or lack of it) which describes the literal _type_. Avoid having to remember whether it's `rb` or `br`. 

This applies to an even greater degree if more prefix modifiers are added, such as `m` for multiline or `f` for formatting. The unified syntax actually reduces the number of different string literal types.

File: compiler/rustc_lexer/src/lib.rs
Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs:200]: > Why is an unequal number of hashes allowed in older editions?

For backwards compatibility. Consider code in the wild with `#"foo"`. Since lexing is mostly edition independent, we have to store some token that can be broken into `#`,`"foo"` for edition 2021 or `#"foo"<invalid>` on edition 2024.

> Wouldn't it make sense for `n_end_hashes` to have the `Option` instead of `n_start_hashes`?

That would make more sense. However, moving the `Option` causes the enum to grow in size, from two bytes to three bytes. Essentially, this is a layout optimization.

I'll make a note to fix that comment to represent this better.

File: compiler/rustc_lexer/src/lib.rs
Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs]: A couple questions:

1. Are you okay with adding an edition field to cursor?
2. Should I include that refactoring in this PR,
3. Or should I follow up with a separate PR for that refactoring?

**Comment by wesleywiser** [comment]
We discussed this in the T-compiler triage meeting today. The consensus was that we strongly prefer not to have snapshotting/unlimited lookahead in the lexer and would rather the design not require that even when accounting for lexing multiple editions. Given how fundamental lexing is to any kind of tool that processes Rust code, not just rustc, keeping it as simple as possible is very important.

**Comment by traviscross** [comment]
We discussed this PR in our edition call this week.

What's the next step here?

@pitaj: Are you waiting for feedback to move forward with making changes, is this waiting on changes you plan to make, or is this waiting on review?

cc @nnethercote

File: compiler/rustc_lexer/src/lib.rs
Comment (joshtriplett) [reply compiler/rustc_lexer/src/lib.rs:200]: > However, moving the Option causes the enum to grow in size, from two bytes to three bytes. Essentially, this is a layout optimization.

@pitaj Would you get the same layout optimization if you put `n_end_hashes` as the first field and then moved the option to that field? Nothing else in the code should care about the field order, so you can reorder the fields for layout optimization.

Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs:200]: Are you suggesting the following? 
```rust
GuardedStr { n_end_hashes: Option<u8>, n_start_hashes: NonZeroU8 }
```
The order doesn't make a difference. The discriminant can't be stored in a separate field.

Comment (joshtriplett) [reply compiler/rustc_lexer/src/lib.rs:200]: @pitaj :facepalm: Sorry, I managed to completely gloss over that one was `NonZeroU8` and the other was just `u8`. Please ignore.

**Comment by traviscross** [comment]
We discussed this in the lang triage call today.

We wanted to acknowledge the concerns that compiler team members have voiced about, as @compiler-errors well put it, "raising the complexity class of the lexer in ways that we've been trying to avoid."

We understand how taking this on to support a reservation is unappealing.

At the same time, many on the call felt we were in fact going to want the feature for which this is saving space.  We wanted to take another look at the open questions on that feature, but we generally felt that the space being reserved is what we'd need for that, and that this feature would raise the same need for edition-dependent lexing.

We noted that, if we don't save this space in Rust 2024, then the feature itself would require a new edition, and the delay imposed by that would plausibly dissipate the energy needed for progress on this.

Some also felt that, regardless, we're likely going to end up with edition-dependent lexing for one feature or another.

We didn't form a consensus here or seek to form one.  We were just reviewing the situation and looking for ways to perhaps unblock it.

We're interested and curious about the thoughts others might have about what to do here, and how we might give some clarity to @pitaj.

**Comment by nnethercote** [comment]
I'm not a fan of the # strings in general. We already have a lot of string literal kinds, with quite a bit of implementation complexity to support them. If we were designing Rust from scratch today then sure, `#` strings might be better than raw strings. But with raw strings already present, the benefits of `#` strings are outweighed by the complexity costs (implementation, specification, education, etc), IMO. There was mention of possibly eventually deprecating raw strings, which arguably helps with the education complexity, but makes the implementation and specification complexity a little worse.

There is an argument that `#` strings unify with vanilla string literals, but I find that a weak argument. Sure, you can say "normal string literals are a degenerate case of `#` strings". But in practice the implementation would be entirely distinct. And from an education aspect they would be mostly distinct -- you'd introduce `#` literals later.

File: compiler/rustc_lexer/src/lib.rs
Comment (oli-obk) [inline compiler/rustc_lexer/src/lib.rs:769]: This comment is out of date

File: compiler/rustc_lexer/src/lib.rs
Comment (oli-obk) [inline compiler/rustc_lexer/src/lib.rs:773]: Instead of creating a backup, you can also just peek one char ahead to see if it is `#` or `"` and return `Pound` in that case. In all other cases, assume it's a double quoted string and parse that. That saves us the infinite lookahead here and leaves it to `rustc_parse::lexer` to handle this in an edition gated way.

Still doesn't make me happy, but the ship on rustc_parse's lexer resetting the lexer with edition rules has already sailed because no one noticed it when that was implemented.

**Review by oli-obk — DISMISSED** [review]

**Comment by oli-obk** [comment]
unassigning myself as I'm going on leave for 4 months starting next week

**Comment by pitaj** [comment]
I have heavily refactored the implementation to move the snapshotting to `rustc_parse::lexer` where snapshotting is also used for `c` strings. I also changed the lexing to only check for a guarded string if `#"`, `##"`, `###"`, or `####` are encountered.

@rustbot ready

**Comment by traviscross** [comment]
@fee1-dead: Perhaps you might be interested to have a look, given that the approach taken here is intended to be similar to the one that you took for C string literals in #113476.

**Review by fee1-dead — APPROVED** [review]
only skimmed over the implementation (don't take this as a full review), but the approach and the testing looks good.

File: compiler/rustc_lint/src/context/diagnostics.rs
Comment (estebank) [inline compiler/rustc_lint/src/context/diagnostics.rs:175]: This can be written like this
```suggestion
        BuiltinLintDiag::ReservedString(suggestion) => {
            lints::ReservedString { suggestion }.decorate_lint(diag);
```

**Comment by traviscross** [comment]
r? @petrochenkov

@petrochenkov, you reviewed #113476 which took a similar approach to the one here.  Do you mind having a look at this?

This is an edition item, and we're trying to finish and mark off the ones that we can.

**Comment by petrochenkov** [comment]
The PR description needs an update.

**Comment by petrochenkov** [comment]
The current implementation still has an infinite lexer lookahead, even if it's called from `report_guarded_str` and not from `advance_token`.

Suppose we don't have to care about backward compatibility or editions.
How are the "guarded string literals" supposed to be lexed in that case?

I don't see how it can be done without the infinite lookahead.
`###############"abc"###############` is a string, and `###############` without the following string is still 15 separate  hashes, but you cannot discern between them without looking arbitrarily far.
This is different from `r#` and friends where you know that it's certainly not just hashes, but rather a string or an error.

**Comment by pitaj** [comment]
@petrochenkov

> The current implementation still has an infinite lexer lookahead, even if it's called from `report_guarded_str` and not from `advance_token`.

The infinite lookahead is only for better diagnostics. The current implementation will issue a hard error (or forward compatibility warning) on any of the following: 

```rust
#"
##"
###
####+
```

There is no case in edition 2024+ that `GuardedStrMaybe` can be broken into individual pounds.

(If there aren't already tests for this I should add some)

> Suppose we don't have to care about backward compatibility or editions.
How are the "guarded string literals" supposed to be lexed in that case?

Like this implementation, they can choose to error on `#"`, `##"`, or any length of repeated `#` greater than or equal to 3. This only requires 3-lookahead.

`###############"abc"###############` is a string, and `###############` without the following string is still 15 separate hashes, but you cannot discern between them without looking arbitrarily far.

This implementation considers the second an invalid guarded string literal.

--------

Technically this implementation doesn't follow either RFC exactly. The lang team should consider the implication of reserving any `###+` for this purpose. From a quick search on GitHub, it doesn't appear to exist in the wild.

**Comment by joshtriplett** [comment]
We reviewed this in today's @rust-lang/lang meeting. Given @pitaj's proposal to fully reserve three or more hashes, which would only need 3-lookahead (not infinite lookahead), we'd like to go forward with this.

We're relying here on the assumption that `###` doesn't occur in the wild; if it does we'd want to re-evaluate this. (If `##` didn't occur in the wild we'd be fine with reserving that too; if ###+ turns up in the wild, we can change it to `####`.)

To be explicit: we're also fine reserving three-or-more hashes in *all* editions if they don't occur in the wild, if that's easier and reduces edition-dependence.

**Comment by traviscross** [comment]
We were a bit unclear in the lang meeting today about the current behavior of the PR.  Specifically, we were unclear whether it was proposing to reserve `###` in Rust 2024 or in all editions.

In testing this branch, I see the following:

```rust
//@ compile-flags: -Zunstable-options
//@ revisions: e2021 e2024
//@ [e2021] edition:2021
//@ [e2024] edition:2024
//@ [e2021] check-pass

macro_rules! m {
    ( _, $($x:tt)* ) => {};
    ( 1, $x:tt ) => {};
    ( 2, $x:tt $y:tt ) => {};
    ( 3, $x:tt $y:tt $z:tt ) => {};
}

fn main() {
    m!(1, #);
    m!(2, ##);
    m!(3, ###);
    m!(_, ####);
    //[e2024]~^ ERROR invalid string literal
}
```

That is, it seems to reserve four or more, and only on Rust 2024.

@pitaj: Some questions:

- Does the impl mean to reserve 3 or more as the comment above suggested?
- Why 3?  If we're doing this over an edition anyway, how much churn would we incur by reserving two or more instead?
- Do we gain anything, in terms of avoiding lookahead, by making a reservation like this in all editions?

**Comment by pitaj** [comment]
> Does the impl mean to reserve 3 or more as the comment above suggested?

Sorry, it's been a moment. Misunderstanding my own code, I fell for the off-by-one error. **I should have said 4.** My apologies for the confusion. I had the amount of lookahead correct, but failed to account for the initial `#`.

> Why ~~3~~ 4? If we're doing this over an edition anyway, how much churn would we incur by reserving two or more instead?

A set of 4 is what you get using the maximum lookahead of 3 currently implemented in the lexer. I'm not opposed to reserving 2+. I can do a code search and see what I find, but I strongly suspect it would be dwarfed by the number of `#"` out there (already affected by the changes here).

> Do we gain anything, in terms of avoiding lookahead, by making a reservation like this in all editions?

I think it would just be converting future-compatibility warnings into hard errors. I don't see how it would improve the lookahead story because we'd still want to provide good diagnostics.

**Comment by pitaj** [comment]
> I'm not opposed to reserving 2+. I can do a code search and see what I find

Not exhaustive by any means, but I looked for a while and couldn't find a single example on GitHub.

**Comment by traviscross** [comment]
@pitaj: Based on that, and how we'd felt about it in discussion, I'd suggest changing this to reserve two or more unprefixed `#` tokens in Rust 2024.

(This is still nominated, and we'll confirm in triage that this sounds right to everyone.)

@petrochenkov: Would this and the comments from @pitaj [above](https://github.com/rust-lang/rust/pull/123951#issuecomment-2289145753) resolve the questions you had [raised](https://github.com/rust-lang/rust/pull/123951#issuecomment-2288932836)?

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in lang triage today and confirmed that we would like to reserve the 2+ `#`s in Rust 2024.

**Comment by petrochenkov** [comment]
Ok, so `##` is always considered a string, similarly to how `r#` is always considered a string, and bare `#######` becomes impossible. That resolves the question.

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:761]: This doesn't look like a common situation and the early return shouldn't be important for performance.
So it would be simpler to wrap the skipping condition around the `buffer_lint` below (and then inline and remove `fn split_guarded_str_maybe`).

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:247]:
```suggestion
                rustc_lexer::TokenKind::GuardedStrMaybe => self.maybe_report_guarded_str(start, str_before),
```
We do not always report a guarded str error/lint here.

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:764]:
```suggestion
        let mut cursor = Cursor::new(str_before);
```
It is definitely a cursor, no?

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:795]:
```suggestion
            let sugg = if !span.from_expansion() {
```

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:812]: There is a double error here, one for reservation and another for unterminated string, one would be enough (with the fatal one as a priority).

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:767]: Style nit: this would look better as a `match`.

File: compiler/rustc_parse/src/lexer/mod.rs
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:780]:
```suggestion
                let unterminated = if !terminated { Some(str_start) } else { None };
```
Redundant condition, the result is only used on 2024 edition anyway.

File: compiler/rustc_lexer/src/lib.rs
Comment (petrochenkov) [inline compiler/rustc_lexer/src/lib.rs:222]: Token lengh like this doesn't seem to be used in other literals.
Perhaps it is redundant and the end position can be calculated without it, using the cursor data?

**Comment by Skgland** [comment]
I was being reminded of https://github.com/rust-lang/rust/pull/118825 by this.

As such I am wondering: Does expanding to two or more hashes rather than three or more  interact properly with code examples?

**Comment by nnethercote** [comment]
I try not to be a negative Nancy, and I rarely weigh in on language design issues. But I want to state clearly that I have a bad feeling about guarded strings, both the idea and the implementation. To echo my comments from above: yet another string literal type, when we already have a lot. All for what I consider a very minor functionality advancement that will have relatively few uses.

Implementation complexity is a particular interest of mine. I spend a lot of time cleaning up code in the compiler, making complex things simpler and more streamlined. I have merged multiple PRs doing exactly that kind of cleanup with the code that handles string literals. It's already more complex than you might expect. And the complexity this PR adds -- including edition-specific special casing of a certain number of leading `#` characters, yuk -- is exactly the kind of thing that makes my spidey senses tingle and that I try to fix or improve. And I believe implementation complexity is a useful feedback mechanism to language design. If something is difficult to implement, it's going to also be difficult to specify, and describe, and educate users on. It will also be a permanent knot of complexity in the language and implementation.

A follow-up question to this paragraph is "are there any changes we can make that would improve the situation?"  But I feel like the answer is "not really". It's all a consequence of guarded string literals being sufficiently different to, and not interacting cleanly with, the existing string literals. If we were designing Rust from scratch it would be a different story.

If I were king of Rust I would just veto the feature entirely.

File: compiler/rustc_lexer/src/lib.rs
Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs:222]: Typically the length of the token is returned inside a `Token` alongside the `TokenKind`. Since this isn't a `TokenKind`, we have to return the token length within it.

File: compiler/rustc_lexer/src/lib.rs
Comment (pitaj) [reply compiler/rustc_lexer/src/lib.rs]: I tried moving this to `LiteralKind`, but it required adding more `unreachable!`s, which is usually a bad sign. I'm going to move it back under `TokenKind`.

File: tests/ui/rust-2024/reserved-guarded-strings-via-macro-2.rs
Comment (pitaj) [reply tests/ui/rust-2024/reserved-guarded-strings-via-macro-2.rs:11]: I did it the same way they handled reserved prefixes in edition 2021. I figure it best to keep things consistent where possible.

**Comment by petrochenkov** [comment]
The implementation looks correct to me now, I won't r+ this though because I disagree with the direction.
I won't block this either because it's only a future-proofing restriction that shouldn't be too harmful, and can hopefully be lifted again some time later.
r? @traviscross

**Comment by traviscross** [comment]
Thanks to all the reviewers here.  I understand this has been a tough one.

This is something that lang wants to reserve for now.  If we decide later to not follow through on this feature direction, we can always unreserve this.  So in that light...

@bors r+

**Comment by traviscross** [comment]
But it looks like there's a conflict that needs to be resolved first, so...

@bors r-

cc @pitaj

**Comment by pitaj** [comment]
That daily looks unrelated. Can we try rerunning bors?

**Comment by ehuss** [comment]
@bors retry

CI error will be fixed by #131461.

**Pull request merged**
